File: core-py/src/oigtl/net/ws_client.py
"""WebSocket client — Python counterpart to a browser or Node WS peer.

Parallel to :class:`~oigtl.net.Client` in shape (connect, send,
receive, dispatch, ``raw_messages``, ``send_raw``, ``run``) but
uses WebSocket framing: one binary frame = one IGTL message.

No WSS / TLS yet.

    async with await WsClient.connect("ws://tracker.lab:18945/") as c:
        await c.send(Transform(matrix=[...]))
        env = await c.receive(Status)
"""

from __future__ import annotations

import asyncio
import contextlib
from datetime import timedelta
from typing import (
    Any,
    AsyncIterator,
    Awaitable,
    Callable,
    TypeVar,
)

import websockets
from pydantic import BaseModel
from websockets.asyncio.client import ClientConnection, connect as ws_connect

from oigtl.codec import RawBody, unpack_message
from oigtl.net._options import (
    ClientOptions,
    Envelope,
    RawMessage,
    as_timedelta,
)
from oigtl.net.errors import (
    ConnectionClosedError,
    FramingError,
    TimeoutError as NetTimeoutError,
)
from oigtl.runtime.exceptions import CrcMismatchError, ProtocolError
from oigtl.runtime.header import HEADER_SIZE, pack_header, unpack_header
from oigtl_corpus_tools.codec.crc64 import crc64

__all__ = ["WsClient"]

M = TypeVar("M", bound=BaseModel)
Handler = Callable[[Envelope[Any]], Awaitable[None]]


class WsClient:
    """Async OpenIGTLink client over WebSocket.

    Same researcher-facing surface as :class:`~oigtl.net.Client` —
    typed send/receive, dispatch decorator, async iterator,
    ``raw_messages`` escape hatch. No resilience features in this
    initial pass; layer on if needed.

    Construct via :meth:`connect` (classmethod, awaits the handshake).
    """

    # --------------------------------------------------------------
    # Construction / teardown
    # --------------------------------------------------------------

    def __init__(
        self,
        *,
        ws: ClientConnection,
        url: str,
        options: ClientOptions,
    ) -> None:
        self._ws = ws
        self._url = url
        self._options = options
        self._send_lock = asyncio.Lock()
        self._closed = asyncio.Event()
        self._run_stop = asyncio.Event()

        self._handlers: dict[str, Handler] = {}
        self._unknown_handler: Handler | None = None
        self._error_handler: Callable[[BaseException], Awaitable[None]] | None = None

    @classmethod
    async def connect(
        cls,
        url: str,
        options: ClientOptions | None = None,
    ) -> "WsClient":
        """Open a WS connection and return a ready :class:`WsClient`.

        Accepts a full URL (``ws://host:port/`` or ``ws://host:port/path``).
        Honours ``options.connect_timeout``. Raises
        :class:`~oigtl.net.errors.TimeoutError` on budget exhaustion,
        :class:`~oigtl.net.errors.ConnectionClosedError` on other
        failures. ``wss://`` not supported yet.
        """
        opt = options or ClientOptions()
        if not (url.startswith("ws://") or url.startswith("wss://")):
            raise ValueError(
                f"ws URL must start with ws:// or wss://; got {url!r}"
            )
        if url.startswith("wss://"):
            raise NotImplementedError(
                "wss:// (TLS) not supported yet; use ws:// for now"
            )

        try:
            coro = ws_connect(
                url,
                max_size=(
                    opt.max_message_size + HEADER_SIZE
                    if opt.max_message_size > 0
                    else None
                ),
            )
            if opt.connect_timeout is None:
                ws = await coro
            else:
                ws = await asyncio.wait_for(
                    coro,
                    timeout=opt.connect_timeout.total_seconds(),
                )
        except asyncio.TimeoutError as e:
            raise NetTimeoutError(
                f"ws connect to {url} timed out after "
                f"{opt.connect_timeout}"
            ) from e
        except (OSError, websockets.InvalidHandshake,
                websockets.InvalidURI) as e:
            raise ConnectionClosedError(
                f"ws connect to {url} failed: {e}"
            ) from e

        return cls(ws=ws, url=url, options=opt)

    # --------------------------------------------------------------
    # Introspection
    # --------------------------------------------------------------

    @property
    def options(self) -> ClientOptions:
        return self._options

    @property
    def url(self) -> str:
        """The URL originally passed to :meth:`connect`."""
        return self._url

    @property
    def is_connected(self) -> bool:
        return not self._closed.is_set()

    # --------------------------------------------------------------
    # Close
    # --------------------------------------------------------------

    async def close(self) -> None:
        """Close the WebSocket. Idempotent."""
        if self._closed.is_set():
            return
        self._closed.set()
        self._run_stop.set()
        with contextlib.suppress(Exception):
            await self._ws.close()

    async def __aenter__(self) -> "WsClient":
        return self

    async def __aexit__(self, *_: object) -> None:
        await self.close()

    # --------------------------------------------------------------
    # Send
    # --------------------------------------------------------------

    async def send(
        self,
        message: BaseModel,
        *,
        device_name: str | None = None,
        timestamp: int = 0,
    ) -> None:
        """Frame and transmit *message* as one WebSocket binary frame."""
        if self._closed.is_set():
            raise ConnectionClosedError("client is closed")

        type_id = getattr(type(message), "TYPE_ID", None)
        if not isinstance(type_id, str):
            raise TypeError(
                f"{type(message).__name__} has no TYPE_ID"
            )
        body = message.pack()
        # v1 framing — see :meth:`oigtl.net.client.Client.send`.
        header = pack_header(
            version=1,
            type_id=type_id,
            device_name=device_name or self._options.default_device,
            timestamp=timestamp,
            body=body,
        )
        await self._send_wire(header + body)

    async def send_raw(self, msg: RawMessage) -> None:
        """Send already-framed wire bytes as one binary frame."""
        if self._closed.is_set():
            raise ConnectionClosedError("client is closed")
        await self._send_wire(msg.wire)

    async def _send_wire(self, wire: bytes) -> None:
        async with self._send_lock:
            try:
                await self._ws.send(wire)
            except websockets.ConnectionClosed as e:
                self._closed.set()
                raise ConnectionClosedError(f"ws send failed: {e}") from e

    # --------------------------------------------------------------
    # Receive
    # --------------------------------------------------------------

    async def receive_any(
        self,
        *,
        timeout: timedelta | float | int | None = None,
    ) -> Envelope[BaseModel]:
        budget = as_timedelta(timeout) or self._options.receive_timeout
        coro = self._receive_one()
        if budget is None:
            return await coro
        try:
            return await asyncio.wait_for(
                coro, timeout=budget.total_seconds(),
            )
        except asyncio.TimeoutError as e:
            raise NetTimeoutError(
                f"receive_any timed out after {budget}"
            ) from e

    async def receive(
        self,
        message_type: type[M],
        *,
        timeout: timedelta | float | int | None = None,
    ) -> Envelope[M]:
        expected = getattr(message_type, "TYPE_ID", None)
        if not isinstance(expected, str):
            raise TypeError(
                f"{message_type.__name__} has no TYPE_ID"
            )
        budget = as_timedelta(timeout) or self._options.receive_timeout
        loop = asyncio.get_running_loop()
        deadline = (
            loop.time() + budget.total_seconds()
            if budget is not None else None
        )

        while True:
            if deadline is not None:
                remaining = deadline - loop.time()
                if remaining <= 0:
                    raise NetTimeoutError(
                        f"receive({message_type.__name__}) timed out "
                        f"after {budget}"
                    )
                try:
                    env = await asyncio.wait_for(
                        self._receive_one(), timeout=remaining,
                    )
                except asyncio.TimeoutError as e:
                    raise NetTimeoutError(
                        f"receive({message_type.__name__}) timed out "
                        f"after {budget}"
                    ) from e
            else:
                env = await self._receive_one()

            if env.header.type_id == expected:
                return env  # type: ignore[return-value]
            await self._dispatch(env)

    async def messages(self) -> AsyncIterator[Envelope[BaseModel]]:
        """Async iterator over every received message, in order."""
        try:
            while not self._closed.is_set():
                yield await self._receive_one()
        except ConnectionClosedError:
            return

    async def raw_messages(self) -> AsyncIterator[RawMessage]:
        """Async iterator yielding each frame as a :class:`RawMessage`."""
        try:
            while not self._closed.is_set():
                yield await self._receive_one_raw()
        except ConnectionClosedError:
            return

    async def _receive_one(self) -> Envelope[BaseModel]:
        raw = await self._receive_one_raw()
        # The framer that produced *raw* already verified the CRC,
        # so skip the second check here.
        body = raw.wire[HEADER_SIZE:]
        return unpack_message(
            raw.header, body, loose=True, verify_crc=False,
        )

    async def _receive_one_raw(self) -> RawMessage:
        if self._closed.is_set():
            raise ConnectionClosedError("client is closed")

        try:
            frame = await self._ws.recv()
        except websockets.ConnectionClosedOK:
            self._closed.set()
            raise ConnectionClosedError("peer closed connection")
        except websockets.ConnectionClosed as e:
            self._closed.set()
            raise ConnectionClosedError(f"ws recv closed: {e}") from e

        if isinstance(frame, str):
            raise FramingError(
                "received text WebSocket frame; OIGTL requires binary"
            )

        data: bytes = bytes(frame)
        if len(data) < HEADER_SIZE:
            raise FramingError(
                f"WS binary frame {len(data)} bytes < header {HEADER_SIZE}"
            )

        try:
            header = unpack_header(data[:HEADER_SIZE])
        except ValueError as e:
            raise ProtocolError(str(e)) from e

        body_size = int(header.body_size)
        expected_size = HEADER_SIZE + body_size
        if len(data) != expected_size:
            raise FramingError(
                f"WS frame size {len(data)} does not match declared "
                f"body_size={body_size} (expected {expected_size})"
            )

        body = data[HEADER_SIZE:]
        computed = crc64(body)
        if computed != header.crc:
            raise CrcMismatchError(
                f"header crc=0x{header.crc:016x} "
                f"body crc=0x{computed:016x}"
            )

        return RawMessage(header=header, wire=data)

    # --------------------------------------------------------------
    # Dispatch
    # --------------------------------------------------------------

    def on(
        self,
        message_type: type[M],
    ) -> Callable[[Handler], Handler]:
        type_id = getattr(message_type, "TYPE_ID", None)
        if not isinstance(type_id, str):
            raise TypeError(
                f"{message_type.__name__} has no TYPE_ID"
            )

        def register(handler: Handler) -> Handler:
            self._handlers[type_id] = handler
            return handler

        return register

    def on_unknown(self, handler: Handler) -> Handler:
        self._unknown_handler = handler
        return handler

    def on_error(
        self,
        handler: Callable[[BaseException], Awaitable[None]],
    ) -> Callable[[BaseException], Awaitable[None]]:
        self._error_handler = handler
        return handler

    async def run(self) -> None:
        """Dispatch loop — read and route until the peer closes."""
        while not self._run_stop.is_set():
            try:
                env = await self._receive_one()
            except ConnectionClosedError:
                return
            except BaseException as e:
                if self._error_handler is not None:
                    await self._error_handler(e)
                    continue
                raise
            try:
                await self._dispatch(env)
            except BaseException as e:
                if self._error_handler is not None:
                    await self._error_handler(e)
                else:
                    raise

    async def _dispatch(self, env: Envelope[Any]) -> None:
        handler = self._handlers.get(env.header.type_id)
        if handler is not None:
            await handler(env)
        elif self._unknown_handler is not None:
            await self._unknown_handler(env)


# Backwards-compatible alias — see the matching note in
# :mod:`oigtl.net.client`. New code should use :class:`oigtl.codec.RawBody`.
_RawBody = RawBody

File: core-ts/src/net/ws_client.ts
/**
 * Async OpenIGTLink client over WebSocket.
 *
 * Parallels the Node-TCP {@link Client} in shape (connect, send,
 * receive, dispatch, messages) but uses WebSocket framing — one
 * binary frame = one IGTL message. No `V3Framer`, no
 * `ByteAccumulator`; the WS layer already delimits messages.
 *
 * Browser-safe: imports nothing from Node. Uses `globalThis.WebSocket`
 * by default; pass `{ webSocket: SomeImpl }` to override (e.g. the
 * `ws` npm package for Node 20/21 where native WebSocket isn't
 * available).
 *
 * No WSS/TLS yet. `wss://` URLs work at runtime (the browser
 * WebSocket API handles TLS), but we haven't tested that path.
 *
 *     import { WsClient } from "@openigtlink/core/net/ws";
 *     import { Transform, Status } from "@openigtlink/core/messages";
 *
 *     const c = await WsClient.connect("ws://tracker.lab:18945/");
 *     await c.send(new Transform({ matrix: [...] }));
 *     const reply = await c.receive(Status);
 *     await c.close();
 */

import { extractContent, unpackMessage } from "../codec.js";
import { type MessageCtor } from "../runtime/dispatch.js";
import { packHeader, unpackHeader, HEADER_SIZE } from "../runtime/header.js";
import { crc64 } from "../runtime/crc64.js";
import { CrcMismatchError } from "../runtime/errors.js";
import {
  ConnectionClosedError,
  FramingError,
  TransportError,
  TransportTimeoutError,
} from "./errors.js";
import { RawBody, type Envelope } from "./envelope.js";
import type {
  ClientOptions,
  ResolvedClientOptions,
} from "./options.js";
import { resolveClientOptions } from "./options.js";
import type {
  Handler,
  SendableCtor,
  SendableMessage,
} from "./types.js";

export type { Handler, SendableCtor, SendableMessage } from "./types.js";

// ---------------------------------------------------------------------------
// Runtime WebSocket shim
// ---------------------------------------------------------------------------

/**
 * Minimal shape we require from a WebSocket implementation.
 * The browser `WebSocket` class and the Node `ws` package's
 * `WebSocket` class both satisfy this — we only use the standard
 * Web-IDL surface.
 */
export interface WebSocketLike {
  readonly readyState: number;
  binaryType: "arraybuffer" | "blob" | "nodebuffer" | string;
  send(data: ArrayBuffer | Uint8Array): void;
  close(code?: number, reason?: string): void;
  addEventListener(
    type: "open" | "close" | "error" | "message",
    listener: (event: any) => void,  // eslint-disable-line @typescript-eslint/no-explicit-any
  ): void;
}

export interface WebSocketLikeCtor {
  new (url: string): WebSocketLike;
  readonly CONNECTING: number;
  readonly OPEN: number;
  readonly CLOSING: number;
  readonly CLOSED: number;
}

/** Options specific to WsClient — extends the shared ClientOptions. */
export interface WsClientOptions extends ClientOptions {
  /**
   * WebSocket constructor to use. Defaults to `globalThis.WebSocket`,
   * which works in browsers and Node 22+. For Node 20/21, install
   * the `ws` npm package and pass `{ webSocket: WebSocket }` where
   * `WebSocket` is `(await import("ws")).WebSocket`.
   */
  webSocket?: WebSocketLikeCtor;
}

function resolveWebSocketCtor(
  override?: WebSocketLikeCtor,
): WebSocketLikeCtor {
  if (override !== undefined) return override;
  const g = globalThis as unknown as {
    WebSocket?: WebSocketLikeCtor;
  };
  if (g.WebSocket !== undefined) return g.WebSocket;
  throw new TransportError(
    "No WebSocket constructor available. Browsers and Node 22+ have " +
      "one natively; for Node 20/21, pass `{ webSocket: ... }` from " +
      "the `ws` npm package.",
  );
}

// ---------------------------------------------------------------------------
// Deferred primitive (shared with TCP Client idiom; duplicated to
// keep the WS module browser-safe — no Node imports).
// ---------------------------------------------------------------------------

interface Deferred<T> {
  readonly promise: Promise<T>;
  resolve(value: T): void;
  reject(err: unknown): void;
}

function deferred<T>(): Deferred<T> {
  let resolve!: (v: T) => void;
  let reject!: (e: unknown) => void;
  const promise = new Promise<T>((res, rej) => {
    resolve = res;
    reject = rej;
  });
  return { promise, resolve, reject };
}

// ---------------------------------------------------------------------------
// WsClient
// ---------------------------------------------------------------------------

interface Incoming {
  header: ReturnType<typeof unpackHeader>;
  body: Uint8Array;
}

export class WsClient {
  private readonly ws: WebSocketLike;
  private readonly _url: string;
  private readonly opts: ResolvedClientOptions;

  private readonly ready: Incoming[] = [];
  private waiting: Deferred<Incoming> | null = null;

  private closed = false;
  private closeCause: Error | null = null;

  private readonly handlers = new Map<string, {
    handler: Handler<unknown>;
    ctor: MessageCtor | undefined;
  }>();
  private unknownHandler: Handler<RawBody> | null = null;
  private errorHandler: ((err: unknown) => void | Promise<void>) | null = null;
  private running = false;

  private constructor(
    ws: WebSocketLike,
    url: string,
    opts: ResolvedClientOptions,
  ) {
    this.ws = ws;
    this._url = url;
    this.opts = opts;
    this.wire();
  }

  // -------------------------------------------------------------
  // Construction
  // -------------------------------------------------------------

  static async connect(
    url: string,
    options?: WsClientOptions,
  ): Promise<WsClient> {
    if (!url.startsWith("ws://") && !url.startsWith("wss://")) {
      throw new Error(
        `ws URL must start with ws:// or wss://; got "${url}"`,
      );
    }
    const opts = resolveClientOptions(options);
    const WS = resolveWebSocketCtor(options?.webSocket);

    const ws = new WS(url);
    // Binary frames must arrive as ArrayBuffer so we can work with
    // them uniformly in browser + Node. "nodebuffer" (Node ws
    // default) gives us Buffer; "arraybuffer" gives us ArrayBuffer.
    ws.binaryType = "arraybuffer";

    await new Promise<void>((resolve, reject) => {
      let settled = false;
      let timer: ReturnType<typeof setTimeout> | undefined;

      const onOpen = () => {
        if (settled) return;
        settled = true;
        if (timer) clearTimeout(timer);
        resolve();
      };
      const onError = (ev: unknown) => {
        if (settled) return;
        settled = true;
        if (timer) clearTimeout(timer);
        try { ws.close(); } catch { /* noop */ }
        reject(
          new ConnectionClosedError(
            `ws connect to ${url} failed: ${errorMessage(ev)}`,
          ),
        );
      };
      const onClose = (ev: unknown) => {
        if (settled) return;
        settled = true;
        if (timer) clearTimeout(timer);
        reject(
          new ConnectionClosedError(
            `ws connect to ${url} failed during handshake: ${
              errorMessage(ev)
            }`,
          ),
        );
      };

      ws.addEventListener("open", onOpen);
      ws.addEventListener("error", onError);
      ws.addEventListener("close", onClose);

      if (opts.connectTimeoutMs !== undefined) {
        timer = setTimeout(() => {
          if (settled) return;
          settled = true;
          try { ws.close(); } catch { /* noop */ }
          reject(
            new TransportTimeoutError(
              `ws connect to ${url} timed out after ${opts.connectTimeoutMs}ms`,
            ),
          );
        }, opts.connectTimeoutMs);
        (timer as { unref?: () => void }).unref?.();
      }
    });

    return new WsClient(ws, url, opts);
  }

  private wire(): void {
    this.ws.addEventListener("message", (ev: { data: unknown }) => {
      try {
        this.onMessage(ev.data);
      } catch (err) {
        this.failWith(err);
      }
    });
    this.ws.addEventListener("close", () => this.onClose(null));
    this.ws.addEventListener("error", (ev: unknown) => {
      this.onClose(
        new ConnectionClosedError(
          `ws error: ${errorMessage(ev)}`,
        ),
      );
    });
  }

  private onMessage(data: unknown): void {
    // Browser: data is ArrayBuffer (binaryType="arraybuffer") or string.
    // Node ws: data is ArrayBuffer because we set binaryType above.
    if (typeof data === "string") {
      throw new FramingError(
        "received text WebSocket frame; OIGTL requires binary",
      );
    }
    let bytes: Uint8Array;
    if (data instanceof ArrayBuffer) {
      bytes = new Uint8Array(data);
    } else if (ArrayBuffer.isView(data)) {
      const v = data as ArrayBufferView;
      bytes = new Uint8Array(v.buffer, v.byteOffset, v.byteLength);
    } else {
      throw new FramingError(
        `unexpected frame payload type: ${Object.prototype.toString.call(data)}`,
      );
    }

    if (bytes.length < HEADER_SIZE) {
      throw new FramingError(
        `WS binary frame ${bytes.length} bytes < header ${HEADER_SIZE}`,
      );
    }
    const header = unpackHeader(bytes.subarray(0, HEADER_SIZE));
    const bodySize = Number(header.bodySize);
    if (!Number.isSafeInteger(bodySize)) {
      throw new FramingError(
        `bodySize ${header.bodySize} exceeds Number.MAX_SAFE_INTEGER`,
      );
    }
    const expected = HEADER_SIZE + bodySize;
    if (bytes.length !== expected) {
      throw new FramingError(
        `WS frame size ${bytes.length} does not match declared ` +
          `bodySize=${bodySize} (expected ${expected})`,
      );
    }
    if (
      this.opts.maxMessageSize > 0 &&
      bodySize > this.opts.maxMessageSize
    ) {
      throw new FramingError(
        `bodySize ${bodySize} exceeds maxMessageSize ${this.opts.maxMessageSize}`,
      );
    }

    const body = bytes.slice(HEADER_SIZE);
    const actual = crc64(body);
    if (actual !== header.crc) {
      throw new CrcMismatchError(header.crc, actual);
    }

    const inc: Incoming = { header, body };
    if (this.waiting !== null) {
      const w = this.waiting;
      this.waiting = null;
      w.resolve(inc);
    } else {
      this.ready.push(inc);
    }
  }

  private onClose(err: Error | null): void {
    if (this.closed) return;
    this.closed = true;
    const cause = err ?? new ConnectionClosedError("peer closed connection");
    this.closeCause = cause;
    if (this.waiting !== null) {
      const w = this.waiting;
      this.waiting = null;
      w.reject(this.closeCause);
    }
  }

  private failWith(err: unknown): void {
    const e =
      err instanceof Error ? err : new ConnectionClosedError(String(err));
    this.closed = true;
    this.closeCause = e;
    try { this.ws.close(); } catch { /* noop */ }
    if (this.waiting !== null) {
      const w = this.waiting;
      this.waiting = null;
      w.reject(e);
    }
  }

  // -------------------------------------------------------------
  // Introspection
  // -------------------------------------------------------------

  get url(): string { return this._url; }
  get options(): ResolvedClientOptions { return this.opts; }
  get isConnected(): boolean { return !this.closed; }

  // -------------------------------------------------------------
  // Close
  // -------------------------------------------------------------

  async close(): Promise<void> {
    if (this.closed) return;
    this.running = false;
    const ended = new Promise<void>((resolve) => {
      this.ws.addEventListener("close", () => resolve());
    });
    try { this.ws.close(); } catch { /* noop */ }
    // Budget the drain so a misbehaving peer doesn't hang close().
    await Promise.race([ended, timeoutPromise(500)]);
  }

  async [Symbol.asyncDispose](): Promise<void> {
    await this.close();
  }

  // -------------------------------------------------------------
  // Send
  // -------------------------------------------------------------

  async send(
    message: SendableMessage,
    opts?: { deviceName?: string; timestamp?: bigint },
  ): Promise<void> {
    if (this.closed) {
      throw this.closeCause ?? new ConnectionClosedError("client is closed");
    }
    const ctor = message.constructor as unknown as
      SendableCtor & { name?: string };
    const typeId = ctor.TYPE_ID;
    if (typeof typeId !== "string" || typeId.length === 0) {
      throw new TypeError(
        `${ctor.name ?? "message"} has no TYPE_ID; not a generated ` +
          `OpenIGTLink message class?`,
      );
    }

    const body = message.pack();
    // v1 framing — see the matching comment in src/net/client.ts.
    const header = packHeader({
      version: 1,
      typeId,
      deviceName: opts?.deviceName ?? this.opts.defaultDevice,
      timestamp: opts?.timestamp ?? 0n,
      body,
    });
    const wire = new Uint8Array(header.length + body.length);
    wire.set(header, 0);
    wire.set(body, header.length);

    try {
      this.ws.send(wire);
    } catch (err) {
      this.failWith(err);
      throw new ConnectionClosedError(`ws send failed: ${errorMessage(err)}`);
    }
  }

  // -------------------------------------------------------------
  // Receive
  // -------------------------------------------------------------

  async receiveAny(opts?: { timeoutMs?: number }): Promise<Envelope<unknown>> {
    const timeout = opts?.timeoutMs ?? this.opts.receiveTimeoutMs;
    return this.decode(await this.nextIncoming(timeout));
  }

  async receive<T>(
    ctor: MessageCtor<T>,
    opts?: { timeoutMs?: number },
  ): Promise<Envelope<T>> {
    const typeId = ctor.TYPE_ID;
    const total = opts?.timeoutMs ?? this.opts.receiveTimeoutMs;
    const deadline = total === undefined ? undefined : Date.now() + total;

    while (true) {
      const remaining = deadline === undefined ? undefined : deadline - Date.now();
      if (remaining !== undefined && remaining <= 0) {
        throw new TransportTimeoutError(
          `receive(${typeId}) timed out after ${total}ms`,
        );
      }
      const inc = await this.nextIncoming(remaining);
      if (inc.header.typeId === typeId) {
        // Strip v2/v3 ext_header + metadata before ctor.unpack;
        // same reasoning as Client.receive.
        const content = extractContent(inc.header, inc.body);
        return { header: inc.header, body: ctor.unpack(content) as T };
      }
      await this.dispatchOne(inc);
    }
  }

  async *messages(): AsyncGenerator<Envelope<unknown>> {
    try {
      while (!this.closed || this.ready.length > 0) {
        yield this.decode(await this.nextIncoming(undefined));
      }
    } catch (err) {
      if (err instanceof ConnectionClosedError) return;
      throw err;
    }
  }

  private async nextIncoming(
    timeoutMs: number | undefined,
  ): Promise<Incoming> {
    if (this.closed && this.ready.length === 0) {
      throw this.closeCause ?? new ConnectionClosedError("client is closed");
    }
    const queued = this.ready.shift();
    if (queued !== undefined) return queued;

    if (this.waiting !== null) {
      throw new TransportError(
        "another receive() is already in flight; serialise your calls",
      );
    }
    this.waiting = deferred<Incoming>();
    if (timeoutMs === undefined) return this.waiting.promise;

    const timer = setTimeout(() => {
      if (this.waiting !== null) {
        const w = this.waiting;
        this.waiting = null;
        w.reject(
          new TransportTimeoutError(`receive timed out after ${timeoutMs}ms`),
        );
      }
    }, timeoutMs);
    (timer as { unref?: () => void }).unref?.();
    try {
      return await this.waiting.promise;
    } finally {
      clearTimeout(timer);
    }
  }

  /** Decode an {@link Incoming} into a typed {@link Envelope}.
   *
   * CRC was already verified by the framer that produced `inc`, so
   * we skip the second check. Unknown type_ids degrade to
   * {@link RawBody} via `loose: true`.
   */
  private decode(inc: Incoming): Envelope<unknown> {
    return unpackMessage(inc.header, inc.body, {
      loose: true,
      verifyCrc: false,
    });
  }

  // -------------------------------------------------------------
  // Dispatch
  // -------------------------------------------------------------

  on<T>(ctor: MessageCtor<T>, handler: Handler<T>): this {
    this.handlers.set(ctor.TYPE_ID, {
      handler: handler as Handler<unknown>,
      ctor: ctor as MessageCtor,
    });
    return this;
  }

  onUnknown(handler: Handler<RawBody>): this {
    this.unknownHandler = handler;
    return this;
  }

  onError(handler: (err: unknown) => void | Promise<void>): this {
    this.errorHandler = handler;
    return this;
  }

  async run(): Promise<void> {
    this.running = true;
    try {
      while (this.running && !this.closed) {
        let inc: Incoming;
        try {
          inc = await this.nextIncoming(undefined);
        } catch (err) {
          if (err instanceof ConnectionClosedError) return;
          if (this.errorHandler) { await this.errorHandler(err); continue; }
          throw err;
        }
        try {
          await this.dispatchOne(inc);
        } catch (err) {
          if (this.errorHandler) {
            await this.errorHandler(err);
          } else {
            throw err;
          }
        }
      }
    } finally {
      this.running = false;
    }
  }

  private async dispatchOne(inc: Incoming): Promise<void> {
    const env = this.decode(inc);
    const entry = this.handlers.get(inc.header.typeId);
    if (entry !== undefined) {
      await entry.handler(env);
      return;
    }
    if (this.unknownHandler !== null) {
      await this.unknownHandler(env as Envelope<RawBody>);
    }
  }
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function errorMessage(ev: unknown): string {
  if (ev instanceof Error) return ev.message;
  if (ev && typeof ev === "object" && "message" in ev) {
    return String((ev as { message: unknown }).message);
  }
  return String(ev);
}

function timeoutPromise(ms: number): Promise<void> {
  return new Promise((resolve) => {
    const t = setTimeout(resolve, ms);
    (t as { unref?: () => void }).unref?.();
  });
}
